### PR TITLE
Convert to ESM

### PR DESCRIPTION
**Pull request by rtritto**
This change is [Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923)
- - -

BREAKING CHANGE

[Convert to ESM guide](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c)

TODO:
- [x]  `yarn build`
- [x]  `yarn lint`
- [x]  `yarn test`
- [x]  `yarn start-dev`
- [x] check/improve _fileURLToPath_
- [x] add _node:_ protocol

Other changes:
- Fixed deprecated Timestamp in `/bson/spec`
- Changed path from _../build-assets.json_ to _./build-assets.json_ in `/lib/middleware.js` (this should be OK ---> to CHECK)
- Replace _lodash_ dependency with _lodash-es_
- Improve check of _config.js_ in `app.js`
- Add _loadConfig_ function in `app.js`

### CODE REVIEW

**Review by BlackthornYugen — COMMENTED** [review]
> Reviewed 13 of 42 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923)* status: 13 of 42 files reviewed, all discussions resolved (waiting on @BlackthornYugen)

**Comment by rtritto** [comment]
Added the guide and changes.
@BlackthornYugen @shakaran maybe further checks are needed.

**Review by BlackthornYugen — COMMENTED** [review]
> Reviewed 3 of 42 files at r1, 4 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923)* status: 14 of 42 files reviewed, 1 unresolved discussion (waiting on @BlackthornYugen and @rtritto)

___
*[`test/lib/bsonSpec.js` line 4 at r2](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NAyrIYeEIXmKMB_-pUV:-NAyrIYeEIXmKMB_-pUW:b-a6aoyc) ([raw file](https://github.com/mongo-express/mongo-express/blob/05af2ac09d9f8bb37dc735f9366e7a4b61349c4d/test/lib/bsonSpec.js#L4)):*
> ```JavaScript
> import mongodb from 'mongodb';
> import {
>   BSONSymbol, Code, DBRef, MaxKey, MinKey, ObjectId, Timestamp,
> ```

@rtritto @shakaran how do you feel about static imports like this?

I tend to avoid them in Java (or at least use them very carefully) because it can be unclear where a method is coming from. I found the more explicit `bson.ObjectId` to be more clear than `ObjectId`. 

`BSONSymbol` and `toBSON`/`toJsonString` are probably fine to static import but the rest might be confusing?

Thoughts?

**Comment by rtritto** [comment]
I followed this [example](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#examples).
Maybe during the start, there is a static optimization on async loading of modules (partial/default import).
For some dependency we must use named exports because namespace default export is removed/deprecated (in _bson_ is deprecated and will be removed).
For our internal files, we can choose to add default export and rename properly/descriptively all keys.

PS: if you want, I can rewrite history with more atomic commits

**Review by BlackthornYugen — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923)* status: 11 of 42 files reviewed, 4 unresolved discussions (waiting on @BlackthornYugen and @rtritto)

___
*[`lib/routes/collection.js` line 5 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCwDBjBNHaOrTVcVN7:-NBCwDBjBNHaOrTVcVN8:b-u9t73o) ([raw file](https://github.com/mongo-express/mongo-express/blob/9aaadb3769fe3288303bce4d9656e9c114679f08/lib/routes/collection.js#L5)):*
> ```JavaScript
> import { EJSON } from 'bson';
> import csv from '../csv.js';
> import { bytesToSize, buildCollectionURL, roughSizeOfObject } from '../utils.js';
> ```

Would it be more clear if we kept these functions within the utils namespace?

```js
import * as utils from '../utils.js`'
```



_Code quote:_
```JavaScript
import { bytesToSize, buildCollectionURL, roughSizeOfObject } from '../utils.js';
```
___
*[`lib/routes/collection.js` line 8 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCvGzCCuUehrhKN8rF:-NBCvGzCCuUehrhKN8rG:bxh9tpu) ([raw file](https://github.com/mongo-express/mongo-express/blob/9aaadb3769fe3288303bce4d9656e9c114679f08/lib/routes/collection.js#L8)):*
> ```JavaScript
> import {
>   parseObjectId, toBSON, toJsonString, toSafeBSON, toString,
> } from '../bson.js';
> ```

I think it'd be more clear if we kept the bson namespace instead of importing all of these to the root.

```js
import * as bson from '../bson.js';
```



_Code quote:_
```JavaScript
import {
  parseObjectId, toBSON, toJsonString, toSafeBSON, toString,
} from '../bson.js';
```
___
*[`lib/routes/collection.js` line 245 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCvZEG57ULEtyRYRyO:-NBCvZEG57ULEtyRYRyP:b-7ba77z) ([raw file](https://github.com/mongo-express/mongo-express/blob/df46a76716ba26b6026b7bb32c2116ce6e6b336c/lib/routes/collection.js#L245)):*
> ```JavaScript
>         docs[i] = items[i];
>         columns.push(Object.keys(items[i]));
>         items[i] = bson.toString(items[i]);
> ```

Consider keeping the bson namespace.

For example, someone reading this would assume we calling routes.toString, having bson made it obvious we are calling code that doesn't come from this class.

**Comment by BlackthornYugen** [comment]
@rtritto it looks like you allowed maintainers to push changes to this branch so I pushed a commit with fewer static imports as a suggestion. Hope that's okay. If you think it's wrong feel free to revert it.

**Comment by rtritto** [comment]
@BlackthornYugen changes seem good.
I completed all TODO points.
A rebase is needed.

**Review by rtritto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923)* status: 3 of 42 files reviewed, 4 unresolved discussions (waiting on @BlackthornYugen and @shakaran)

___
*[`lib/routes/collection.js` line 5 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCwDBjBNHaOrTVcVN7:-NBEXDPYD5ao07Iw4hqG:b-896fix) ([raw file](https://github.com/mongo-express/mongo-express/blob/9aaadb3769fe3288303bce4d9656e9c114679f08/lib/routes/collection.js#L5)):*
<details><summary><i>Previously, BlackthornYugen (John Steel) wrote…</i></summary><blockquote>

Would it be more clear if we kept these functions within the utils namespace?

```js
import * as utils from '../utils.js`'
```
</blockquote></details>

Done.
___
*[`lib/routes/collection.js` line 8 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCvGzCCuUehrhKN8rF:-NBEX3YFCxRqB836UVYq:b-896fix) ([raw file](https://github.com/mongo-express/mongo-express/blob/9aaadb3769fe3288303bce4d9656e9c114679f08/lib/routes/collection.js#L8)):*
<details><summary><i>Previously, BlackthornYugen (John Steel) wrote…</i></summary><blockquote>

I think it'd be more clear if we kept the bson namespace instead of importing all of these to the root.

```js
import * as bson from '../bson.js';
```
</blockquote></details>

Done.
___
*[`lib/routes/collection.js` line 245 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCvZEG57ULEtyRYRyO:-NBEX7FaATk1koOiJEFm:b-896fix) ([raw file](https://github.com/mongo-express/mongo-express/blob/df46a76716ba26b6026b7bb32c2116ce6e6b336c/lib/routes/collection.js#L245)):*
<details><summary><i>Previously, BlackthornYugen (John Steel) wrote…</i></summary><blockquote>

Consider keeping the bson namespace.

For example, someone reading this would assume we calling routes.toString, having bson made it obvious we are calling code that doesn't come from this class. 
</blockquote></details>

Done.
___
*[`test/lib/bsonSpec.js` line 4 at r2](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NAyrIYeEIXmKMB_-pUV:-NBEX9ptEHTxdvtg3rWd:b-896fix) ([raw file](https://github.com/mongo-express/mongo-express/blob/05af2ac09d9f8bb37dc735f9366e7a4b61349c4d/test/lib/bsonSpec.js#L4)):*
<details><summary><i>Previously, BlackthornYugen (John Steel) wrote…</i></summary><blockquote>

@rtritto @shakaran how do you feel about static imports like this?

I tend to avoid them in Java (or at least use them very carefully) because it can be unclear where a method is coming from. I found the more explicit `bson.ObjectId` to be more clear than `ObjectId`. 

`BSONSymbol` and `toBSON`/`toJsonString` are probably fine to static import but the rest might be confusing?

Thoughts? 
</blockquote></details>

Done.

**Review by shakaran — COMMENTED** [review]
:lgtm:

> Reviewed 5 of 42 files at r1, 2 of 9 files at r2, 26 of 35 files at r6, 1 of 9 files at r9, 5 of 5 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923#-:-NBGTwIx0NbatU5Hqv_6:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @BlackthornYugen and @shakaran)

File: lib/middleware.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { fileURLToPath } from 'node:url';
 import express from 'express';
 import swig from 'swig-templates';
 import fileUpload from 'express-fileupload';
-import swigFilters from './filters.js';
+import * as swigFilters from './filters.js';
```
Comment (BlackthornYugen) [inline lib/middleware.js:6]: Oh! I didn't realize that `import * as <namespace> from 'string'` can be simplified to `import <namespace> from 'string'`. Maybe it's nicer to use the second one?

Comment (rtritto) [reply lib/middleware.js:6]: To reduce bundle size is better to don't use default export.
`import <namespace> from 'string'` can only be used with default export (I removed it in that commit).

**Review by BlackthornYugen — APPROVED** [review]
:lgtm:

> Reviewed 1 of 35 files at r6, 1 of 9 files at r9, 1 of 3 files at r11, 2 of 11 files at r13, 5 of 5 files at r14, 4 of 4 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923#-:-NBMaNPuCC3NZQsXDJ_0:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @rtritto)

___
*[`lib/middleware.js` line 6 at r12](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NAuLFslF7eBzwMzLkAi-r12-6:-NBM_DTF5_fYjPTD1ZgH:bb6un1j) ([raw file](https://github.com/mongo-express/mongo-express/blob/f49c38f29136f4f9f41a32c19674f78739b6ade2/lib/middleware.js#L6)):*
<details><summary><i>Previously, rtritto wrote…</i></summary><blockquote>

To reduce bundle size is better to don't use default export.
`import <namespace> from 'string'` can only be used with default export (I removed it in that commit).
</blockquote></details>

Done
___
*[`lib/routes/collection.js` line 8 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCvGzCCuUehrhKN8rF:-NBM_miA00Lq8PfjClMT:bb6un1j) ([raw file](https://github.com/mongo-express/mongo-express/blob/9aaadb3769fe3288303bce4d9656e9c114679f08/lib/routes/collection.js#L8)):*
<details><summary><i>Previously, rtritto wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done
___
*[`lib/routes/collection.js` line 245 at r3](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NBCvZEG57ULEtyRYRyO:-NBMa0dQ49E5UcSi588a:bb6un1j) ([raw file](https://github.com/mongo-express/mongo-express/blob/df46a76716ba26b6026b7bb32c2116ce6e6b336c/lib/routes/collection.js#L245)):*
<details><summary><i>Previously, rtritto wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done
___
*[`test/lib/bsonSpec.js` line 4 at r2](https://reviewable.io/reviews/mongo-express/mongo-express/923#-NAyrIYeEIXmKMB_-pUV:-NBMaHRQ-8UpfhZsFZad:bb6un1j) ([raw file](https://github.com/mongo-express/mongo-express/blob/05af2ac09d9f8bb37dc735f9366e7a4b61349c4d/test/lib/bsonSpec.js#L4)):*
<details><summary><i>Previously, rtritto wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done

**Review by BlackthornYugen — APPROVED** [review]
> Reviewed 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mongo-express/mongo-express/923)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @rtritto)

**Comment by BlackthornYugen** [comment]
I've built an image from this branch ( ghcr.io/blackthornyugen/mongo-express-docker:v1.0.1-rc ) with ( https://github.com/BlackthornYugen/mongo-express-docker/runs/8255241979 ). Looks good to me. Looks like this is ready to merge?

**Comment by rtritto** [comment]
Yes, it's ready.

**Pull request merged**